src/mpl/src/gavl/mpl_gavl.c

```diff
@@ -321,138 +384,104 @@ int MPL_gavl_tree_free(MPL_gavl_tree_t gavl_tree)
-                        STACK_PUSH(node_stack, inorder_node);
-                        inorder_node = inorder_node->left;
-                    }
-
-                    if (inorder_node->parent != cur_node) {
-                        if (inorder_node->right)
-                            inorder_node->right->parent = inorder_node->parent;
-                        inorder_node->parent->left = inorder_node->right;
-                    } else {
-                        /* only right child of deleted node */
-                        cur_node->right = NULL;
-                    }
-                    val = (void *) cur_node->val;
-                    cur_node->addr = inorder_node->addr;
-                    cur_node->len = inorder_node->len;
-                    cur_node->val = inorder_node->val;
-                    MPL_free(inorder_node);
-                }
-
-                STACK_POP(node_stack, cur_node);
-
-              stack_recovery:
-                gavl_update_node_info(cur_node);
-
-                int lheight = cur_node->left == NULL ? 0 : cur_node->left->height;
-                int rheight = cur_node->right == NULL ? 0 : cur_node->right->height;
-                if (lheight - rheight > 1) {
-                    gavl_tree_node_s *lnode = cur_node->left;
-                    int llheight = lnode->left == NULL ? 0 : lnode->left->height;
-                    if (llheight + 1 == lheight)
-                        gavl_right_rotation(cur_node, lnode);
-                    else
-                        gavl_left_right_rotation(cur_node, lnode);
-                } else if (rheight - lheight > 1) {
-                    gavl_tree_node_s *rnode = cur_node->right;
-                    int rlheight = rnode->left == NULL ? 0 : rnode->left->height;
-                    if (rlheight + 1 == rheight)
-                        gavl_right_left_rotation(cur_node, rnode);
-                    else
-                        gavl_left_rotation(cur_node, rnode);
-                }
-
-                if (!STACK_EMPTY(node_stack)) {
-                    STACK_POP(node_stack, cur_node);
-                    goto stack_recovery;
-                } else {
-                    break;
-                }
-            } while (1);
-
-            while (gavl_tree_iptr->root && gavl_tree_iptr->root->parent != NULL)
-                gavl_tree_iptr->root = gavl_tree_iptr->root->parent;
-
-            if (val && gavl_tree_iptr->gavl_free_fn)
-                gavl_tree_iptr->gavl_free_fn(val);
+            /* search and return the node to be deleted */
+            gavl_tree_search_node(tree_ptr, (uintptr_t) addr, len, INTERSECTION_SEARCH, &dnode);
+            if (dnode == NULL)
+                goto fn_exit;
+
+            /* delete the matched node and return corresponding buffer object val */
+            gavl_tree_delete_node(tree_ptr, dnode, &val);
+
+            /* rebalace tree */
+            gavl_tree_rebalance(tree_ptr);
+
```

Comment:
Can we rebalance the tree only once after deleting all nodes?

Reply:
no, we have to rebalance the tree whenever a node is deleted

Reply:
I understand that we need to maintain the tree balanced when returning to user. However, why do we have to do it during internal deletion? If we don't do it, would `gavl_tree_search_internal` not work correctly at the next iteration (assuming there are multiple intersected nodes)?

Reply:
if you keep deletion without rebalance, all the trace stored in the stack would be meaningless. There will be no way to guarantee this tree as an avl tree anymore; that is we cannot assure that left and right child height difference is at most 1.

Reply:
We want to make sure that left and right hight gap is at most 1 for fast search. However, during a deletion, no one else is searching the tree (assume there is a per-tree critical section). I do not see why we have to spend time to rebalance after every internal deletion. 

Besides, "all the trace stored in the stack would be meaningless" is not the reason why we need rebalance after every internal deletion.

Reply:
We cannot rebalance an avl tree whose left and right child height difference is above 1. This is the reason why we cannot delete all the nodes first and then rebalance the tree.

Reply:
@KaimingOuyang Let me try explain from my guess (since I haven't check the algorithm itself) and correct me if I am off --
I believe the "rebalance" is a very local and optimized operation, ie, knowing the left height and right height is off by 1 and simply rotates the children by 1. So this is efficient and meant to be exercised at each insertion and deletion. If we don't do this *local* balance each time, then since the information on the balancing get lost, the rebalace will be a whole tree rebalance, which will be extremely expensive, more expensive than consecutive local rebalances.

Reply:
actually there is no way to do the whole tree rebalance.

Reply:
> actually there is no way to do the whole tree rebalance.

There is always way 😄 You meant to say there is no efficient way

Reply:
alright, in that case, you are right, very inefficient way to rebalance the whole tree.
I don't consider it as a way, so I say "no way" :)

Reply:
Can you add the explanation in the code comment? It might be better if you can cover these questions:
1. Why do we have to rebalance after every internal deletion
2. What is the expectation of `gavl_tree_rebalance` (i.e., assumes that the height difference is at most 1)
3. Internal logic of `gavl_tree_rebalance`

2 and 3 might be better to be placed near function `gavl_tree_rebalance`.

Reply:
I add more comments as you require, please check.